### PR TITLE
add support for migration to non-upper resourcepool

### LINKED ISSUE #114: Allow migration of upper resourcepool to non-upper resourcepool
We should update the quota of the crq-pointer (i.e., the upper resourcepool of the toNS).

/kind feature

### PR DESCRIPTION
**Pull request by mzeevi**
Fixes #114. With this PR, resources move to the crq-pointer namespace so this also covers migration to non-upper resourcepools. In addition, controller-runtime is bumped to v0.16.3

### CODE REVIEW

**Review by liadlauber — CHANGES_REQUESTED** [review]
Well done! Overall looks good to me. Consider also adding test cases for migration to non-upper resourcepool.
Keep pushing for excellence!

File: internals/controllers/migrationhierarchy/migrationhierarchy_controller.go [hunk 1]
```diff
@@ -85,27 +85,59 @@ func (r *MigrationHierarchyReconciler) reconcile(mhObject *utils.ObjectContext)
 	currentNamespace := mhObject.Object.(*danav1.MigrationHierarchy).Spec.CurrentNamespace
 	toNamespace := mhObject.Object.(*danav1.MigrationHierarchy).Spec.ToNamespace
 
-	ns, err := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Namespace: "", Name: currentNamespace}, &corev1.Namespace{})
-	if err != nil {
-		return ctrl.Result{}, fmt.Errorf("failed getting namespace object '%s': "+err.Error(), currentNamespace)
+	ns, er := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Namespace: "", Name: currentNamespace}, &corev1.Namespace{})
```
[the remark this answers is not in the document]
Comment (mzeevi) [reply internals/controllers/migrationhierarchy/migrationhierarchy_controller.go]: done

File: internals/controllers/migrationhierarchy/migrationhierarchy_controller.go [hunk 2]
```diff
@@ -85,27 +85,59 @@ func (r *MigrationHierarchyReconciler) reconcile(mhObject *utils.ObjectContext)
...
+		return ctrl.Result{}, fmt.Errorf("failed getting quota object '%s': "+er.Error(), currentNamespace)
 	}
+
 	sourceResources := utils.GetQuotaObjectSpec(sourceQuotaObj.Object)
 	rootNSName := utils.GetNSDisplayNameSlice(ns)[0]
 
 	// add the resources that are allocated to the migrated subnamespace to the new parent using UpdateQuota API
 	sourceQuotaObjExists, _, _ := utils.DoesSNSQuotaObjectExist(oldSNS)
 
+	toNS, er := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Namespace: "", Name: toNamespace}, &corev1.Namespace{})
```
[the remark this answers is not in the document]
Comment (mzeevi) [reply internals/controllers/migrationhierarchy/migrationhierarchy_controller.go]: done

File: internals/controllers/migrationhierarchy/migrationhierarchy_controller.go [hunk 3]
```diff
@@ -85,27 +85,59 @@ func (r *MigrationHierarchyReconciler) reconcile(mhObject *utils.ObjectContext)
...
+	if er != nil {
+		err := r.updateMHStatus(mhObject, danav1.Error, er.Error())
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed updating the status of object '%s': "+err.Error(), mhObject.GetName())
+		}
+		return ctrl.Result{}, fmt.Errorf("failed to get namespace '%s': "+er.Error(), toNamespace)
+	}
+
+	toSNSParentName := utils.GetNamespaceParent(toNS.Object)
+	toSNS, er := utils.NewObjectContext(ctx, r.Client, client.ObjectKey{Name: toNamespace, Namespace: toSNSParentName}, &danav1.Subnamespace{})
```
[the remark this answers is not in the document]
Comment (mzeevi) [reply internals/controllers/migrationhierarchy/migrationhierarchy_controller.go:132]: this is consisent right now in the code, we can look into changing that in the future

**Comment by mzeevi** [comment]
I'll open an issue about the tests; I've made a new commit, PTAL

**Review by liadlauber — APPROVED** [review]

**Pull request merged**